[JsonPath] Test against official compliance test suite

mttsch opened the pull request:
Close #60669

| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | Fix https://github.com/symfony/symfony/issues/60669
| License       | MIT

<!--
Replace this notice by a description of your feature/bugfix.
This will help reviewers and should be a good start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too).
 - Features and deprecations must be submitted against the latest branch.
 - For new features, provide some code snippets to help understand usage.
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
 - Never break backward compatibility (see https://symfony.com/bc).
-->

mttsch committed:
[JsonPath] Test against official compliance test suite

stof reviewed (commented):
We should record the info about which commit/release of the CTS we vendor, with a script we can run to fetch the latest version, updating both cts.json and the version.

Otherwise, it will be hard to manage updates when the compliance testsuite adds new tests to increase coverage (which might happen)

mttsch commented:
@stof As this is only one JSON file that can be also be downloaded manually without any issue, I assumed that providing the exact hash based link, see https://github.com/symfony/symfony/pull/60695/files#diff-8312c65e1b644a04bd87f0dee0894ab6712b296af12a1e191d1faaf5a1f52784R37, was sufficient.

stof commented on src/Symfony/Component/JsonPath/Tests/JsonPathComplianceTestSuiteTest.php:
```diff
@@ -0,0 +1,52 @@
+            $this->expectException(JsonCrawlerException::class);
+        }
+
+        $result = (new JsonCrawler(json_encode($document)))->find($selector);
```
I suggest separating the JsonCrawler instantiation from the `find` call, so that invalid selectors don't allow us to expect an exception in the instantiation part. This would be done by making the call `$this->expectException(JsonCrawlerException::class);` after instantiation but before running the selector.

mttsch replied:
Done.

stof commented:
I think we might need a constant in the test containing a list of test names we don't support yet, for which the test would get skipped in CI (maybe with an env variable to force running them when working on them locally).
This would allow us to merge that PR, preventing regressions on tests that we already pass and moving little by little into running it all (removing tests from the skip-list when we fix the issues we have) instead of being able to merge this PR only once we are 100% compliant.

mttsch commented:
@stof I added a makefile and added a const to exclude unsupported files. 

> (maybe with an env variable to force running them when working on them locally)

Not sure about this part (also not done) because it can basically also be done by temporarily commenting out/removing the array contents similar to

> removing tests from the skip-list when we fix the issues we have

mttsch commented:
@stof PR now targets 7.3.

alexandre-daubois reviewed (approved):
A lot of improvements for the components during the last days and this PR will be really helpful to iterate on the crawler. Thank you @mttsch!

nicolas-grekas commented:
Thank you @mttsch.
